src/aws_client.erl

```diff
@@ -1,34 +1,135 @@
+                                       region     => Region,
+                                       endpoint   => Endpoint}}),
+      setup_update_callback(Expiry, Ref, Role),
+      {reply, {ok, Ref}, State}
+  end;
+handle_call({add_creds, {static, AccessKey, SecretKey}, Opts}, _From, State) ->
+  case ets:lookup(creds_dedup, {static, AccessKey}) of
+    [{_, Ref}] -> {reply, {ok, Ref}, State};
+    [] ->
+      {Region, Endpoint} = parse_opts(Opts),
+      Ref = make_ref(),
+      true = ets:insert(creds_dedup, {{static, AccessKey}, Ref}),
+      true = ets:insert(creds, {Ref, #{access_key => AccessKey,
+                                       secret_key => SecretKey,
+                                       region     => Region,
+                                       endpoint   => Endpoint}}),
+      {reply, {ok, Ref}, State}
+  end;
+handle_call(Args, _From, State) ->
+  io:format("Unknown args ~p~n", [Args]),
+  {noreply, State}.
+
+handle_cast(_Message, State) ->
+  {noreply, State}.
+
+handle_info({refresh_metadata, Ref, Role}, State) ->
+  {ok, AccessKey, SecretKey, Expiry} = get_metadata_creds(Role),
+  Creds = get_creds(Ref),
+  true = ets:insert(creds, {Ref, Creds#{access_key => AccessKey,
+                                        secret_key => SecretKey}}),
+  setup_update_callback(Expiry, Ref, Role),
+  {noreply, State};
+
+handle_info(_Message, State) ->
+  {noreply, State}.
+
+code_change(_Prev, State, _Extra) ->
+  {ok, State}.
+
+%%====================================================================
+%% Helper funcs
+%%====================================================================
+parse_opts(Opts) ->
+  Region = proplists:get_value(region, Opts, <<"us-east-1">>),
+  Endpoint = proplists:get_value(endpoint, Opts, <<"amazonaws.com">>),
+  {Region, Endpoint}.
+
+
+get_role() ->
+  {ok, 200, _, ClientRef} = hackney:get(?CREDENTIAL_URL),
+  hackney:body(ClientRef).
+
 
--ifdef(TEST).
--include_lib("eunit/include/eunit.hrl").
+get_metadata_creds(Role) ->
+  {ok, 200, _, ClientRef} = hackney:get([?CREDENTIAL_URL, Role]),
+  {ok, Body} = hackney:body(ClientRef),
+  Map = jsx:decode(Body, [return_maps]),
+  {ok, maps:get(<<"AccessKeyId">>, Map),
+       maps:get(<<"SecretAccessKey">>, Map),
+       maps:get(<<"Expiration">>, Map)}.
 
-%% make_client/0 returns a clienturation map with default values.
-make_client_test() ->
-    ?assertEqual(#{access_key_id => <<"access-key-id">>,
-                   secret_access_key => <<"secret-access-key">>,
-                   region => <<"region">>,
-                   endpoint => <<"amazonaws.com">>,
-                   service => undefined},
-                 make_client(<<"access-key-id">>, <<"secret-access-key">>,
-                             <<"region">>)).
+setup_update_callback(Timestamp, Ref, Role) ->
+  AlertAt = seconds_until_timestamp(Timestamp) - ?ALERT_BEFORE_EXPIRY,
+  erlang:send_after(AlertAt, ?MODULE, {refresh_metadata, Ref, Role}).
```

Comment:
You're better off using `self()` rather than `?MODULE` here. Otherwise, you may get in the fun situation where the gen_server fails, loses the name, gets replaced by the supervisor, gets the name again. Someone reads the metadata credentials and they get added. This sets up a timeout.

Now when the old timeout triggers, you auto-renew it and end up with two concurrent set of time outs for the same cred set.

As time goes, you can end up spamming AWS over failures because the registered name of the process outlives that process and the data it realistically holds in its tables. If you instead use a Pid as a target, and that process crashes, you won't end up sending invalid messages to its replacement.